patches/avsc+5.7.9.patch

```diff
@@ -0,0 +1,3778 @@
++  Type.addError = (err) => { err.fieldName = err.fieldName ? schema.name + '.' + err.fieldName : schema.name; saveAddError(err); };
+   var name = schema.name;
+   if (typeof name != 'string' || !utils.isValidName(name)) {
+-    throw new Error(f('invalid field name: %s', name));
++    Type.addError(new Error(f('invalid field name: %s', name)));
+   }
+ 
+   this.name = name;
+@@ -2816,12 +2818,15 @@ function Field(schema, opts) {
+       case 'ignore':
+         return 0;
+       default:
+-        throw new Error(f('invalid order: %j', order));
++        Type.addError(new Error(f('invalid order: %j', order)));
+     }
+   })(schema.order === undefined ? 'ascending' : schema.order);
+ 
+   var value = schema['default'];
+-  if (value !== undefined) {
++
++  if (value === null && this.type.typeName === 'record') {
++    Type.addError(new Error(f('invalid default for field %s: null is not a %s', this.name, this.type.typeName)));
++  } else if (value !== undefined) {
+     // We need to convert defaults back to a valid format (unions are
+     // disallowed in default definitions, only the first type of each union is
+     // allowed instead).
+@@ -2838,7 +2843,7 @@ function Field(schema, opts) {
+           type.types[0]
+         );
+       }
+-      throw new Error(msg);
++      Type.addError(new Error(msg));
+     }
+     // The clone call above will throw an error if the default is invalid.
+     if (isPrimitive(type.typeName) && type.typeName !== 'bytes') {
+@@ -2848,6 +2853,7 @@ function Field(schema, opts) {
+       this.defaultValue = function () { return type._copy(val); };
+     }
+   }
++  Type.addError = saveAddError;
+ 
+   Object.freeze(this);
+ }
+@@ -2906,7 +2912,7 @@ function Hash() {
+ function readValue(type, tap, resolver, lazy) {
+   if (resolver) {
+     if (resolver._readerType !== type) {
+-      throw new Error('invalid resolver');
++      Type.addError(new Error('invalid resolver'));
+     }
+     return resolver._read(tap, lazy);
+   } else {
+@@ -3038,7 +3044,7 @@ function isJsonBuffer(obj) {
+  * with a hook since the path is not propagated (for efficiency reasons).
+  */
+ function throwInvalidError(val, type) {
+-  throw new Error(f('invalid %j: %j', type.schema(), val));
++  Type.addError(new Error(f('invalid %j: %j', type.schema(), val)));
+ }
+ 
+ function maybeQualify(name, ns) {
+@@ -3294,6 +3300,9 @@ function combineObjects(types, opts) {
+   return Type.forSchema(schema, opts);
+ }
+ 
++const errorsCollector = [];
++
++Type.addError = (err) => errorsCollector.push(err);
+ 
+ module.exports = {
+   Type: Type,
+@@ -3313,5 +3322,6 @@ module.exports = {
+       types[getClassName(typeName)] = TYPES[typeName];
+     }
+     return types;
+-  })()
++  })(),
++  errorsCollector
+ };
+diff --git a/node_modules/avsc/lib/types.js.backup b/node_modules/avsc/lib/types.js.backup
+new file mode 100644
+index 0000000..dc7c69c
+--- /dev/null
++++ b/node_modules/avsc/lib/types.js.backup
```

Comment:
Do we need this backup in the patch?

Reply:
good catch that's a leftover of when i generated the patch i'll remove!

Reply:
Removed thanks!